### PR TITLE
Avoid SIGPIPE signals when connections break.

### LINKED ISSUE #2654: SIGPIPE raised when server disconnects during a download.
If the server closes a connection during a download (or probably during any request, but this is the most obvious case) the application may get a `SIGPIPE` signal.

The issue is that we disable "signal handling" in libcurl, to avoid getting `SIGALARM` in multi-threaded applications, but the downside is that libcurl also disables masking the `SIGPIPE` signal:

https://curl.haxx.se/mail/lib-2010-11/0008.html

The recommended practice is to ignore the signal during the calls to `curl_easy_perform()` and presumably also `curl_multi_perform()` too.

### PR DESCRIPTION
**Pull request by coryan**
Automatically setup a signal handler for SIGPIPE on platforms where it
is needed. libcurl recommends setting CURLOPT_NOSIGNAL for
multi-threaded applications (and we expect the library to be used in
such applications):

https://curl.haxx.se/libcurl/c/threadsafe.html

"When using multiple threads you should set the CURLOPT_NOSIGNAL option
to 1L for all handles."

"Note that setting CURLOPT_NOSIGNAL to 0L will not work in a threaded
situation as there will be race where libcurl risks restoring the
former signal handler while another thread should still ignore it."

But also note that when setting CURLOPT_NOSIGNAL to 1L the documentation
recommends:

"When CURLOPT_NOSIGNAL is set to 1L, your application needs to deal with
the risk of a SIGPIPE (that at least the OpenSSL backend can trigger)."

This change implements this last recommendation.

This fixes #2654.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/2659)

### CODE REVIEW

File: google/cloud/storage/internal/curl_wrappers_disable_sigpipe_handler_test.cc [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+namespace {
+
+extern "C" void test_handler(int) {}
+
+/// @test Verify that installing the libraries
```
Comment (devbww) [inline google/cloud/storage/internal/curl_wrappers_disable_sigpipe_handler_test.cc]: Truncated sentence?

File: google/cloud/storage/internal/curl_wrappers_disable_sigpipe_handler_test.cc [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  return;  // nothing to do
+#elif LIBCURL_VERSION_NUM > 0x072900
+  // libcurl <= 7.29.0 installs its own signal handler for SIGPIPE during
+  // curl_global_init(). Unfortunately 7.29.0 is the default on CentOS-7, and
+  // the tests here fails. We simply skip the test with this ancient library.
+  auto initial_handler = std::signal(SIGPIPE, &test_handler);
+  CurlInitializeOnce(ClientOptions(oauth2::CreateAnonymousCredentials())
+                         .set_enable_sigpipe_handler(false));
+  auto actual = std::signal(SIGPIPE, initial_handler);
+  EXPECT_NE(actual, SIG_IGN);
```
Comment (devbww) [inline google/cloud/storage/internal/curl_wrappers_disable_sigpipe_handler_test.cc]: This expectations looks superfluous given the next one.

File: google/cloud/storage/internal/curl_wrappers_enable_sigpipe_handler_test.cc [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+namespace {
+
+extern "C" void test_handler(int) {}
+
+/// @test Verify that installing the libraries
```
Comment (devbww) [inline google/cloud/storage/internal/curl_wrappers_enable_sigpipe_handler_test.cc]: Truncated sentence?

File: google/cloud/storage/internal/curl_wrappers_enable_sigpipe_handler_test.cc [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+TEST(CurlWrappers, SigpipeHandlerEnabledTest) {
+#if !defined(SIGPIPE)
+  return;  // nothing to do
+#else
+  auto initial_handler = std::signal(SIGPIPE, &test_handler);
+  CurlInitializeOnce(ClientOptions(oauth2::CreateAnonymousCredentials())
+                         .set_enable_sigpipe_handler(true));
+  auto actual = std::signal(SIGPIPE, initial_handler);
+  EXPECT_EQ(actual, SIG_IGN);
+  EXPECT_NE(actual, &test_handler);
```
Comment (devbww) [inline google/cloud/storage/internal/curl_wrappers_enable_sigpipe_handler_test.cc]: This expectation looks superfluous given the previous one.

File: google/cloud/storage/internal/curl_wrappers_enable_sigpipe_handler_test.cc [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+#endif  // defined(SIGPIPE)
```
Comment (devbww) [inline google/cloud/storage/internal/curl_wrappers_enable_sigpipe_handler_test.cc]: Should we re-call CurlInitializeOnce() and expect that it leaves the initial_handler() in place?

File: google/cloud/storage/tests/slow_reader_integration_test.cc [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+  // Construct a large object, or at least large enough that it is not
+  // downloaded in the first chunk.
+  auto large_text = CreateLargeText(16 * 1024 * 1024);
+
+  // Create an object with the contents to download.
+  StatusOr<ObjectMetadata> source_meta = client->InsertObject(
+      bucket_name, object_name, large_text, IfGenerationMatch(0));
+  ASSERT_STATUS_OK(source_meta);
+
+  // Create a iostream to read the object back, when running against the
```
Comment (devbww) [inline google/cloud/storage/tests/slow_reader_integration_test.cc]: s/a/an/

File: google/cloud/storage/tests/slow_reader_integration_test.cc [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  std::cout << "Reading " << std::flush;
+  while (!stream.eof()) {
+    std::cout << ' ' << slow_reader_period.count() << "s" << std::flush;
+    std::this_thread::sleep_for(slow_reader_period);
+    stream.read(buffer.data(), size);
+    if (!stream.status().ok()) {
+      // TODO(#2655) - automatically restart the download.
+      // Until recently, the code crashed before this point, so mark this as a
+      // "success", though obviously (#2655) we want to do better.
+      std::cout << " DONE (" << stream.status() << ")\n" << std::flush;
```
Comment (devbww) [inline google/cloud/storage/tests/slow_reader_integration_test.cc]: s/ << std::flush//

File: google/cloud/storage/tests/slow_reader_integration_test.cc [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+      auto status = client->DeleteObject(bucket_name, object_name);
+      EXPECT_STATUS_OK(status);
+      return;
+    }
+    EXPECT_STATUS_OK(stream.status());
+    if (slow_reader_period < max_slow_reader_period) {
+      slow_reader_period += std::chrono::minutes(1);
+    }
+  }
+  std::cout << " DONE\n" << std::flush;
```
Comment (devbww) [inline google/cloud/storage/tests/slow_reader_integration_test.cc]: s/ << std::flush//

Comment (coryan) [reply google/cloud/storage/tests/slow_reader_integration_test.cc]: Done.

File: google/cloud/storage/tests/slow_reader_integration_test.cc (hunk 7 above)
Comment (coryan) [reply google/cloud/storage/tests/slow_reader_integration_test.cc]: Done.

File: google/cloud/storage/tests/slow_reader_integration_test.cc (hunk 6 above)
Comment (coryan) [reply google/cloud/storage/tests/slow_reader_integration_test.cc]: Fixed.

File: google/cloud/storage/internal/curl_wrappers_enable_sigpipe_handler_test.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_wrappers_enable_sigpipe_handler_test.cc]: Done.

File: google/cloud/storage/internal/curl_wrappers_enable_sigpipe_handler_test.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_wrappers_enable_sigpipe_handler_test.cc]: Removed.

File: google/cloud/storage/internal/curl_wrappers_enable_sigpipe_handler_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_wrappers_enable_sigpipe_handler_test.cc]: Fixed.

File: google/cloud/storage/internal/curl_wrappers_disable_sigpipe_handler_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_wrappers_disable_sigpipe_handler_test.cc]: Removed.

File: google/cloud/storage/internal/curl_wrappers_disable_sigpipe_handler_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_wrappers_disable_sigpipe_handler_test.cc]: Fixed.

**Comment by coryan** [comment]
PTAL.

**Pull request merged**